java/arcs/core/storage/DirectStoreMuxer.kt

```diff
@@ -83,48 +120,66 @@ class DirectStoreMuxer<Data : CrdtData, Op : CrdtOperation, T>(
     }
 
     /**
-     * Sends the provided [ProxyMessage] to the store backing the provided [referenceId].
+     * Sends the provided [ProxyMessage] to the store backing the provided [muxId].
      *
-     * A new store will be created for the [referenceId], if necessary.
+     * A new store will be created for the [muxId], if necessary.
      */
     suspend fun onProxyMessage(
-        message: ProxyMessage<Data, Op, T>,
-        referenceId: String
+        message: ProxyMessage<Data, Op, T>
     ): Boolean {
-        val (id, store) = store(referenceId)
-        val deMuxedMessage: ProxyMessage<Data, Op, T> = when (message) {
-            is SyncRequest -> SyncRequest(id)
-            is ModelUpdate -> ModelUpdate(message.model, id)
-            is Operations -> if (message.operations.isNotEmpty()) {
-                Operations(message.operations, id)
-            } else return true
+        val muxId = requireNotNull(message.muxId) {
+            "messages sent to Direct Store Muxer must have a muxId"
         }
-        return store.onProxyMessage(deMuxedMessage)
+        // registerToStore will only register to the store if the id is not already registered
+        registerToStore(message.id!!, muxId)
+        val (_, store) = store(muxId)
+        return store.onProxyMessage(message)
     }
 
-    /* internal */ suspend fun setupStore(referenceId: String): StoreRecord<Data, Op, T> {
+    /* internal */ suspend fun setupStore(muxId: String): StoreRecord<Data, Op, T> {
         val store = DirectStore.create<Data, Op, T>(
             StoreOptions(
-                storageKey = storageKey.childKeyWithComponent(referenceId),
+                storageKey = storageKey.childKeyWithComponent(muxId),
                 type = backingType
             )
         )
 
-        val id = store.on(callbackFactory(referenceId))
+        // Return a new Record with initially empty idSet
+        return StoreRecord(mutableSetOf<Int>(), store)
+    }
 
-        // Return a new Record and add it to our local stores, keyed by muxId.
-        return StoreRecord(id, store)
+    private suspend fun registerToStore(
+        id: Int,
+        muxId: String
+    ) {
+        // The Direct Store Muxer can register several callbacks to the same store.
+        // When the Direct Store Muxer receives a message from a Direct Store, it utilizes the
+        // message id (determined by which callback the Direct Store invoked) to determine which
+        // observer to redirect the message to.
+        val (idSet, store) = store(muxId)
+        storeMutex.withLock {
+            if (!idSet.contains(id)) {
+                store.on(
+                    ProxyCallback { message ->
+                        proxyManager.getCallback(id)?.invoke(message.withMuxId(muxId))
+                    },
+                    id
+                )
+                idSet.add(id)
+                callbackIdToMuxIdMap.getOrPut(id, { mutableSetOf() }).add(muxId)
```

Comment:
This mutation is happening outside of the other lock (synchronized block on proxyCallbacks). I think we may need to consider overhauling the threadsafety mechanisms we've got in this class. 

I'm a bit worried we're going to run into hard-to-debug concurrency issues with the state of things.

Reply:
I've moved all accesses to / mutations of callbackIdToMuxIdMap to inside a synchronized block. In this code block in particular I've moved the three calls (`store.on(..`, `idSet.add(..`, and `callbackIdToMuxIdMap.getOrPut(..`) to inside the synchronized block, is that the right thing to do?

As for overhauling the threadsafety mechanisms we have in this class, are you suggesting we make changes in this PR or in a future PR? And what changes are you thinking of? The callbackIdToMuxIdMap is used to quickly unregister an id from stores in the `off` function. Maybe we can move this logic into the proxyCallbackManager.